modules/ggl-tls-helper/bin/ggl-tls-helper.c

```diff
@@ -0,0 +1,626 @@
+        if (ret != GGL_ERR_OK) {
+            return 1;
+        }
+        using_proxy = true;
+        const char *port = (parse_port != NULL) ? parse_port : "80";
+        ret = create_tcp_connection(parse_host, port, &tcp_fd);
+
+        OPENSSL_free(parse_host);
+        OPENSSL_free(parse_port);
+    } else {
+        ret = create_tcp_connection(arg_endpoint, arg_port, &tcp_fd);
+    }
+    if (ret != GGL_ERR_OK) {
+        GGL_LOGE("Failed to create TCP connection.");
+        return 1;
+    }
+
+    BIO *bio = BIO_new_socket(tcp_fd, BIO_NOCLOSE);
+    if (bio == NULL) {
+        GGL_LOGE("Failed to create openssl BIO.");
+        ERR_print_errors_cb(openssl_err_cb, NULL);
+        return 1;
+    }
+
+    if (using_proxy) {
+        ret = http_proxy_connect(bio, arg_endpoint, arg_port);
+        if (ret != GGL_ERR_OK) {
+            return 1;
+        }
+    }
+
+    SSL *ssl = NULL;
+    ret = tls_handshake(arg_endpoint, bio, &ssl);
+    if (ret != GGL_ERR_OK) {
+        return 1;
+    }
+    GGL_CLEANUP(cleanup_ssl, ssl);
+
+    GGL_LOGI("Connection established successfully.");
+
+    bool using_ktls = ktls_enabled(ssl);
+
+    int parent_fds[2];
+
+    if (using_ktls) {
+        GGL_LOGD("Using kernel TLS offload.");
+        parent_fds[0] = tcp_fd;
+        parent_fds[1] = -1;
+    } else {
+        int sys_ret = socketpair(AF_UNIX, SOCK_STREAM, 0, parent_fds);
+        if (sys_ret == -1) {
+            GGL_LOGE("Failed to create socketpair: %m.");
+            return 1;
+        }
+    }
+
+    ret = send_socket_to_parent(parent_fds[0]);
+    if (ret != GGL_ERR_OK) {
+        return 1;
+    }
+
+    if (using_ktls) {
+        GGL_LOGD("Transferred kTLS socket; exiting.");
+        return 0;
+    }
+
+    GGL_LOGD("Preparing to proxy socket traffic.");
+
+    // Close the parent's end of the socketpair
+    (void) close(parent_fds[0]);
+    int proxy_fd = parent_fds[1];
+
+    // Proxy data between socketpair and TLS connection
+    struct pollfd fds[2] = { { .fd = proxy_fd, .events = POLLIN },
+                             { .fd = tcp_fd, .events = POLLIN } };
+
+    uint8_t buffer[4096];
+
+    while (fds[0].fd >= 0 || fds[1].fd >= 0) {
+        int poll_ret = poll(fds, 2, -1);
+        if (poll_ret < 0) {
+            GGL_LOGE("Poll error: %m.");
+            return 1;
+        }
+
+        // Data from proxy socket to TLS
+        if (fds[0].revents & POLLIN) {
+            GglBuffer buf = GGL_BUF(buffer);
+            do {
+                ret = ggl_file_read_partial(proxy_fd, &buf);
+            } while (ret == GGL_ERR_RETRY);
+
+            if (ret == GGL_ERR_NODATA) {
+                GGL_LOGD("Proxy socket closed.");
+                int shutdown_ret = SSL_shutdown(ssl);
+                if (shutdown_ret < 0) {
+                    GGL_LOGE("Openssl shutdown on network socket failed.");
+                    ERR_print_errors_cb(openssl_err_cb, NULL);
+                }
+                (void) shutdown(tcp_fd, SHUT_WR);
```

Comment:
Nit: maybe use `fds[1]`?
I mean maybe pick one - we are using `fds[]` and `tcp_fd`/`proxy_fd` interchangeably.

Reply:
fds[1] is the handle in the poll which we use for checking the poll and disabling the slot by setting it to -1. tcp_fd will remain valid even after we clear the poll slot.